infrastructure/src/main/java/org/corfudb/infrastructure/remotecorfutable/DatabaseHandler.java

```diff
@@ -0,0 +1,472 @@
+                throw e;
+            }
+            //otherwise, the database is empty
+            return new LinkedList<>();
+        }  else {
+            //begin at first value -> extract prefix and add timestamp to it
+            byte[] startingPrefix = KeyEncodingUtil.extractEncodedKey(iter.key());
+            byte[] startingKey = KeyEncodingUtil.constructDatabaseKey(startingPrefix, timestamp);
+            iter.close();
+            iterOptions.close();
+            return scanInternal(startingKey,numEntries,streamID, false);
+        }
+    }
+
+    /**
+     * This function performs a cursor scan of the database, returning 20 entries (if possible).
+     * @param encodedKeyBegin The starting point of the scan.
+     * @param streamID The stream backing the database to scan.
+     * @return A list of results from the scan.
+     * @throws RocksDBException An error occuring in iteration.
+     */
+    public List<byte[][]> scan(byte[] encodedKeyBegin, ByteString streamID) throws RocksDBException, DatabaseOperationException {
+        return scan(encodedKeyBegin, 20, streamID);
+    }
+
+    /**
+     * This function provides an interface for a cursor scan of the database.
+     * @param encodedKeyBegin The starting point of the scan.
+     * @param numEntries The number of results requested.
+     * @param streamID The stream backing the database to scan.
+     * @return A list of results from the scan.
+     * @throws RocksDBException An error occuring in iteration.
+     */
+    public List<byte[][]> scan(byte[] encodedKeyBegin, int numEntries, ByteString streamID) throws RocksDBException, DatabaseOperationException {
+        return scanInternal(encodedKeyBegin, numEntries, streamID, true);
+    }
+
+    private List<byte[][]> scanInternal(byte[] encodedKeyBegin, int numEntries, ByteString streamID, boolean skipFirst) throws RocksDBException, DatabaseOperationException {
+        if (!columnFamilies.containsKey(streamID)) {
+            throw new DatabaseOperationException("SCAN", "Invalid stream ID");
+        }
+        KeyEncodingUtil.VersionedKey start = KeyEncodingUtil.extractVersionedKey(encodedKeyBegin);
+        long timestamp = start.getTimestamp();
+        ReadOptions iterOptions = new ReadOptions().setTotalOrderSeek(true);
+        RocksIterator iter = database.newIterator(columnFamilies.get(streamID), iterOptions);
+        byte[] prevPrefix = start.getEncodedRemoteCorfuTableKey();
+        byte[] currPrefix;
+        boolean encounteredPrefix = false;
+        iter.seek(encodedKeyBegin);
+        List<byte[][]> results = new LinkedList<>();
+        int elements = 0;
+        while (iter.isValid()) {
+            currPrefix = KeyEncodingUtil.extractEncodedKey(iter.key());
+            if (Arrays.equals(currPrefix, prevPrefix)) {
+                if (!encounteredPrefix) {
+                    if (!skipFirst) {
+                        results.add(new byte[][]{iter.key(), iter.value()});
```

Comment:
agree with codacy

Reply:
I'll find some way to reduce the complexity of the nested if-else, I'll keep as a behavior placeholder right now and add TODO.